File: donghwa/arxiv/palindrome/2384_largest_palindromic_number.cpp
#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

class Solution {
public:
    string largestPalindromic(string num) {
        vector<int> cnt(10);
        for (char c: num) {
            cnt[c - '0']++;
        }
        
        string lp, rp; // left and right partition
        for (int i: num) {
            for (int j = 9; j >= 0; j--)  {
                if (cnt[j] > 1 && (j > 0 || lp.size())) {
                    lp += '0' + j;
                    rp += '0' + j;
                    cnt[j] -= 2;
                    break;
                }
            }
        }
        for (int i = 9; i >= 0; i--) {
            if (cnt[i]) {
                lp += '0' + i; break;
            }
        }
        reverse(begin(rp), end(rp));
        return lp + rp;
    }
};
int main() {
    string num1 = "444947137";
    Solution solution;

    string result1 = solution.largestPalindromic(num1);
    cout << "Example 1 Output: " << result1 << endl;  // Output: "7449447"

    string num2 = "00009";
    string result2 = solution.largestPalindromic(num2);
    cout << "Example 2 Output: " << result2 << endl;  // Output: "9"

    return 0;
}

File: donghwa/arxiv/palindrome/2384_largest_palindromic_number.py

def largestPalindrome(num: str) -> str:

    ans = []
    b = [str(x) for x in range(9, -1, -1)]
    from collections import defaultdict

    a = defaultdict(int)

    for x in num:
        a[x] += 1

    for x in b:
        n = len(ans)
        if n % 2 == 0:
            if a[x] > 0:
                ans = ans[:n // 2] + [x] * a[x] + ans[n // 2:]
        else:
            if x == '0':
                if len(ans) != 1:
                    ans = ans[:n // 2] + [x] * (a[x] // 2) + [ans[n // 2]] + [x] * (a[x] // 2) + ans[n // 2 + 1:]
            else:
                if a[x] >= 2:
                    ans = ans[:n // 2] + [x] * (a[x] // 2) + [ans[n // 2]] + [x] * (a[x] // 2) + ans[n // 2 + 1:]

    res = "".join(ans)
    return str(int(res))

# Test cases
num1 = "444947137"
print("Example 1 Output:", largestPalindrome(num1))  # Output: "7449447"

num2 = "00009"
print("Example 2 Output:", largestPalindrome(num2))  # Output: "9"
